### PR TITLE
refactor: refactor preprocessing

### PR DESCRIPTION
**Pull request by slettner**
This PR adds:
- usage of `volumes` to mount weights into the executor
- build-in preprocessing

### CODE REVIEW

File: encoder.py [hunk 1]
```diff
@@ -136,12 +139,17 @@ def _get_docs_batch_generator(self, docs: DocumentArray, parameters: Dict):
     def _compute_embeddings(self, docs_batch_generator: Iterable) -> None:
         with torch.no_grad():
             for document_batch in docs_batch_generator:
-                blob_batch = np.stack([d.blob for d in document_batch])
-                images = self._maybe_move_channel_axis(blob_batch)
+                blob_batch = [d.blob for d in document_batch]
+                if self.use_default_preprocessing:
```
Comment (JoanFM) [inline encoder.py:143]: I am not so sure these transformations can work in batches.

File: encoder.py [hunk 2]
```diff
@@ -78,14 +75,23 @@ def __init__(
...
+        model = getattr(models, self.model_name)(pretrained=True)
 
         self.model = self._extract_feature_from_torch_module(model)
         self.model.to(torch.device(self.device))
+        if self.pool_strategy is not None:
+            self.pool_fn = getattr(np, self.pool_strategy)
+
+        self._preprocess = T.Compose([
+            T.ToPILImage(),
+            T.Resize(256),
```
Comment (JoanFM) [inline encoder.py:87]: Is this even needed?

Comment (slettner) [reply encoder.py:87]: Yes, had to add it

File: encoder.py (hunk 1 above)
Comment (slettner) [reply encoder.py:143]: They are applied per image
```python
def _preprocess_image(self, images: List[np.array]) -> List[np.ndarray]:
    return [self._preprocess(img) for img in images]
```

**Comment by jakobkruse1** [comment]
After merging, do not forget to update the hub component.

**Comment by slettner** [comment]
@jakob1111996 Already done. But the bump to jina2.0 is missing. I would do it in another PR

File: encoder.py [hunk 3]
```diff
@@ -107,22 +113,19 @@ def _get_pooling(self, feature_map: 'np.ndarray') -> 'np.ndarray':
         return self.pool_fn(feature_map, axis=(2, 3))
 
     @requests
-    def encode(self, docs: Optional[DocumentArray], parameters: Dict, **kwargs):
+    def encode(self, docs: Optional[DocumentArray], parameters: Optional[Dict] = None, **kwargs):
```
Comment (JoanFM) [inline encoder.py]: parameters is not Optional

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: encoder.py (hunk 3 above)
Comment (slettner) [reply encoder.py]: Fixed in https://github.com/jina-ai/executor-image-torch-encoder/pull/4/commits/99541b46497b695f24e3db23495950d9fe614b9f

**Review by JoanFM — APPROVED** [review]

**Pull request merged**